Dialog/SpritesEditorDialog.cpp

```diff
@@ -0,0 +1,561 @@
+                                                   "According to the vanilla game engine, the previous loaded palettes and tiles can be overwritten by the others'\n"
+                                                   "So set the load table carefully.\n"
+                                                   "The editor use different palettes for every entity, but ingame, some of them share the same palettes and tiles\n"
+                                                   "you will find the editor act differently from the real game play.\n"
+                                                   "In this case, you need to modify all of the entities who sharing the same tiles and palettes."));
+}
+
+/// <summary>
+/// Deconstructor of SpritesEditorDialog class.
+/// </summary>
+SpritesEditorDialog::~SpritesEditorDialog()
+{
+    delete ui;
+}
+
+void SpritesEditorDialog::SetSelectedEntityColorId(int colorID)
+{
+    if (colorID > 15) colorID = 15;
+    curEntityColorIdInPalette = colorID;
+    SelectionBox_Color->setPos(colorID * 8, curEntityPalId * 8);
+    SelectionBox_Color->setVisible(true);
+    QColor color = GetCurEntityPtr()->GetPalette(curEntityPalId)[colorID];
+    ui->label_CurColorValue->setText(QString("RGB888: (") +
+                                    QString::number(color.red(), 10) + QString(", ") +
+                                    QString::number(color.green(), 10) + QString(", ") +
+                                    QString::number(color.blue(), 10) + QString(") RGB555: (") +
+                                    QString::number(color.red() >> 3, 10) + QString(", ") +
+                                    QString::number(color.green() >> 3, 10) + QString(", ") +
+                                    QString::number(color.blue() >> 3, 10) + QString(")"));
+}
+
+/// <summary>
+/// Reset sprite palette Id
+/// </summary>
+/// <param name="paletteID">
+/// New palette id value.
+/// </param>
+void SpritesEditorDialog::SetSelectedEntityPaletteId(int paletteID)
+{
+    int palnum = GetCurEntityPtr()->GetPalNum();
+    if (paletteID >= palnum) paletteID = palnum - 1;
+    curEntityPalId = paletteID;
+    ui->label_PalID->setText(QString::number(paletteID));
+    RenderSpritesTileMap();
+    SetSelectedSpriteTile(0);
+}
+
+/// <summary>
+/// Reset sprite color in one specified palette.
+/// </summary>
+/// <param name="paletteId">
+/// palette id value to find color to set.
+/// </param>
+/// <param name="colorId">
+/// color id value to find color to set.
+/// </param>
+void SpritesEditorDialog::SetColor(int paletteId, int colorId)
+{
+    int palnum = GetCurEntityPtr()->GetPalNum();
+    if (paletteId >= palnum) paletteId = palnum - 1;
+    if (colorId > 15) colorId = 15;
+    QColor color = QColorDialog::getColor(Qt::black, this);
+    color.setAlpha(0xFF);
+    if(color.isValid())
```

Comment:
Why do we have this check?

Reply:
because you can cancel the QColorDialog without selecting some color